### PR TITLE
Render tables as gtable

### LINKED ISSUE #180: Feature request: combine gt with cowplot or patchwork
Following #155, it would be fantastic if gt would play nicely with [cowplot](https://github.com/wilkelab/cowplot), [patchwork](https://github.com/thomasp85/patchwork), or similar packages. A few graphics + neatly formatted summary table is a powerful combination for communicating insights effectively!

### LINKED ISSUE #961: Convert gt to grob for cowplot + patchwork
## Prework

- [x] Read and abide by **gt**'s [code of conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/) and [contributing guidelines](https://github.com/rstudio/gt/blob/master/.github/CONTRIBUTING.md).
- [x] Search for duplicates among the [existing issues](https://github.com/rstudio/gt/issues) (both open and closed).

## Proposal

In some situations it would be helpful to add a table into the layout of some graphics. While this is possible in RMarkdown/Quarto/HTML/etc, some users prefer `patchwork` or `cowplot` when aggregating their plots for other tools.

This would add a dependency for `grid` (already there from `ggplot2`) and `png`.

This is related to https://github.com/rstudio/gt/issues/420 and https://github.com/rstudio/gt/issues/180 - but per Rich, opening as a new issue/FR.

``` r
library(patchwork)
library(ggplot2)
library(cowplot)
library(gt)

gt_tab <- head(mtcars) |> 
  gt()

test_plot <- mtcars |> 
  ggplot(aes(x = disp, y = mpg)) +
  geom_point() +
  labs(
    alt = "a plot with text"
  )

gt_grob <- function(gt_object, ...){
  
  out_name <- file.path(
    tempfile(pattern = "file", tmpdir = tempdir(), fileext = ".png")
  )
  
  gtsave(gt_object, out_name, ...)
  
  in_png <- png::readPNG(out_name)
  
  on.exit(file.remove(out_name), add=TRUE)
  
  grid::rasterGrob(in_png)
  
}

gt_plot <- gt_grob(gt_tab)

combo_plot <- test_plot +
  wrap_elements(gt_plot)

cow_plot <- plot_grid(test_plot, gt_plot)

ggsave("patchwork.png", combo_plot, height = 3.5, width = 10)
ggsave("cowplot.png", cow_plot, height = 3.5, width = 10)
```

``` r
### Patchwork works
knitr::include_graphics("patchwork.png")
```

``` r

### Cowplot works
knitr::include_graphics("cowplot.png")
```

<sup>Created on 2022-06-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### PR DESCRIPTION
**Pull request by teunbrand**
# Summary

This PR aims to fix #180 and fix #961, and conflicts with #1015.
Briefly, it adds a `render_as_gtable()` function that builds the table as a gtable. 

# Related GitHub Issues and PRs

- Ref: #180, #961, #1015

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct.html).
- [ ] I have listed any major changes in the [NEWS](https://github.com/rstudio/gt/blob/master/NEWS.md).
- [x] I have added [`testthat`](https://github.com/r-lib/testthat) unit tests to [`tests/testthat`](https://github.com/rstudio/gt/tree/master/tests/testthat) for any new functionality.

# Description

I was asked by @thomasp85 whether I could take a stab at #180, so that patchwork and other (gg)plot composition packages could integrate nice tables that {gt} offers. This PR adds a method for rendering tables in the grid graphics system. It consists of two important parts.

1. Getting a layout of the table.
2. Rendering the cells.

## Table layout

For {gtable}, we need to know the layout of the table in advance. For every cell in the table, we need to know the left/right/top/bottom cell in the gtable it occupies, similar to the rowspan and colspan attributes determine in html how wide or tall a cell is. I'm not proficient enough in html to parse the table layout from the html that the `render_as_html()` function produces. Instead, the layout is computed separately from, but analogous to, the html code. All the important layout code is captured in the `create_{part}_component_g()` functions ('g' for grid or gtable), that mirror the `create_{part}_component_h()` ('h' for html).

## Rendering cells

For rendering cells, I've opted for following the css classes and styles that are used in the html output. These are parsed and translated to graphical parameters that {grid} understands. Mostly, this requires a bunch of parsers and a way to consolidate different classes/styles, but it works well enough for most cases. Of course, not all html/css features can be translated to grid, like `overflow-x`, collapsing borders, automatic text wrapping etc. These are thus not implemented. In addition, if the cell text itself contains html, this is not natively understood by grid. To make this slightly less of a pain, `render_as_gtable()` has a `text_grob` argument that can take e.g. `gridtext::richtext_grob()` to render the html in the cells better than vanilla grid would.

# Current state

I think I'm finished with all the internal cogs and gears, there are just a few dangling ends that still need to be taken care of:

* The vertical spacing seems a bit off compared to the html. @thomasp85, should we take extra care around font ascents and descents in any way?
* How should we expose this to users? Currently, `render_as_gtable()`, like `render_as_html()` is an internal function.

Here is a small demo what it looks like:

``` r
devtools::load_all("~/packages/gt")
#> ℹ Loading gt

table <- pizzaplace |>
  dplyr::filter(type %in% c("classic", "veggie")) |>
  dplyr::group_by(type, size) |>
  dplyr::summarize(
    sold = dplyr::n(),
    income = sum(price),
    .groups = "drop"
  ) |>
  gt(rowname_col = "size", groupname_col = "type") |>
  tab_header(title = "Pizzas Sold in 2015") |>
  fmt_integer(columns = sold) |>
  fmt_currency(columns = income) |>
  summary_rows(
    fns = list(label = "All Sizes", fn = "sum"),
    side = c("top"),
    fmt = list(
      ~ fmt_integer(., columns = sold),
      ~ fmt_currency(., columns = income)
    )
  ) |>
  tab_options(
    summary_row.background.color = "gray95",
    row_group.as_column = TRUE
  ) |>
  tab_stub_indent(
    rows = everything(),
    indent = 2
  ) |>
  grand_summary_rows(
    columns = c("sold", "income"),
    fns = list(Sum ~ sum(.)),
    fmt = ~ fmt_number(.)
  ) |>
  tab_caption("Here be caption text") |>
  tab_spanner(
    label = "Spanner",
    columns = c("sold", "income")
  ) |>
  tab_stubhead("Stubhead label") |>
  tab_source_note("Source: the pizzaria") |>
  tab_footnote("Pineapples not included")

gtable <- render_as_gtable(table)
plot(gtable)
```

<sup>Created on 2024-02-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: tests/testthat/test-utils_render_grid.R [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+  gt <- gt(df)
+  test <- create_heading_component_g(prep_data(gt))
+  expect_null(test)
+
+  test <- create_heading_component_g(prep_data(tab_header(gt, title = "Foo")))
+
+  expect_equal(dim(test), c(1, 8))
+  expect_equal(test$right, 2)
+  expect_equal(test$label, "Foo")
+  expect_in("gt_title", test$classes[[1]])
```
Comment (olivroy) [inline tests/testthat/test-utils_render_grid.R:43]:
```suggestion
  expect_contains(test$classes[[1]], "gt_title")
```

File: tests/testthat/test-utils_render_grid.R [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+
+  test <- create_heading_component_g(prep_data(tab_header(gt, "Foo", "Bar")))
+  expect_equal(dim(test), c(2, 8))
+  expect_equal(test$top, c(1, 2))
+  expect_equal(test$label, c("Foo", "Bar"))
+  expect_in("gt_subtitle", test$classes[[2]])
```
Comment (olivroy) [inline tests/testthat/test-utils_render_grid.R:49]:
```suggestion
  expect_contains(test$classes[[2]], "gt_subtitle")
```

File: tests/testthat/test-utils_render_grid.R [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+  gt <- gt %>% tab_header("Foo", "Bar") %>%
+    tab_footnote("x", cells_title("title")) %>%
+    tab_footnote("y", cells_title("subtitle"))
+
+  test <- create_heading_component_g(prep_data(gt))
+  expect_equal(dim(test), c(2, 8))
+  expect_equal(
+    grepl("gt_footnote_marks", test$label),
+    c(TRUE, TRUE)
+  )
```
Comment (olivroy) [inline tests/testthat/test-utils_render_grid.R:60]:
```suggestion
  expect_match(test$label, "gt_footnote_marks")
```

File: tests/testthat/test-utils_render_grid.R [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+  expect_null(test)
+
+  test <- gt %>%
+    summary_rows(
+      fns = list("min", "max")
+    ) %>%
+    prep_data() %>%
+    summary_rows_g(group_id = c("X", "Y"))
+
+  expect_equal(length(test), 2)
```
Comment (olivroy) [inline tests/testthat/test-utils_render_grid.R:224]:
```suggestion
  expect_length(test, 2)
```

**Review by olivroy — COMMENTED** [review]
This looks exciting! I will test this, here are some minor suggestions for tests.

**Comment by rich-iannone** [comment]
@teunbrand , this is excellent! Thank you so much for taking this on.

**Comment by rich-iannone** [comment]
Because we're still testing older R versions, we're stuck with using `%>%` in our tests.

**Comment by teunbrand** [comment]
Oops, sorry my bad 😅

**Comment by rich-iannone** [comment]
@teunbrand do you feel this is ready for a final review (i.e., no more work on your end for the time being)?

**Comment by teunbrand** [comment]
I think this is almost complete, the only open question is about how this should face the user.
All is working as I think it should, save for a few rare cases where also the html layout is off (for example I ran into #1561 at some point). So yeah, I'm satisfied with the bulk of it, so I think this is ready to be reviewed.

**Comment by rich-iannone** [comment]
Somewhat strangely, tests on R 3.6.3 aren't passing because of something new. Looks like a problem both on the gt side and possibly with gtable (and maybe stemming from the gt column widths problem?). I will investigate this in a short while.

**Comment by teunbrand** [comment]
Part of the issue might be the old v1 units {grid} had before R 4.0.0, back when `unit(10, "pt") + unit(10, "pt")` was stored as a delayed operation `sum(unit(10, "pt"), unit(10, "pt"))` instead of `unit(20, "pt")`. I've skipped these tests on old R versions for now, but it shouldn't affect the layout of the table, it is just much harder to do `unit()` tests. If {gt} follows tidyverse R version support rules, this should cease being an issue in April.

**Review by rich-iannone — APPROVED** [review]
LGTM!

**Pull request merged**